### DISCUSSION TITLE
Sponsor block causes a mess

### DISCUSSION THREAD

**Discussion by christoph-heinrich** [tomasklaen/uosc#182] [General]
I've tried out sponsor block because of #178 and noticed that it causes quite a mess with the chapters.
The video https://www.youtube.com/watch?v=MGe1GCVZTAA ends up with the following chapters (formatted as chapter.time chapter.title):
```
0 tech news shall endure
8 Wi-Fi 7 demoed at 5Gbps
73 Buy Your Mom An iPhone
133 Intel Arc specs, RTX 40 event
183 Manscaped Ultra Premium Collection
183.014 Sponsor segment start (37a943)
220 QUICK BITS
220.353 Selfpromo segment start (6e74de)
220.458 Sponsor segment end (37a943)
226 A16 Bionic leaked benchmarks
226.272 Selfpromo segment end (6e74de)
258 FSR 2.1
286 YouTube Player for Education
317 Crystal Dynamics gets Tomb Raider back
350 Runway AI video editing
386.302 Outro segment start (b79e09)
396.04 Outro segment end (b79e09)
```
The generated chapter ranges are a bit better with:
``` diff
diff --git a/scripts/uosc.lua b/scripts/uosc.lua
index e12399b..a729bdd 100644
--- a/scripts/uosc.lua
+++ b/scripts/uosc.lua
@@ -3184,7 +3184,6 @@ for _, definition in ipairs(split(options.chapter_ranges, ' *,+ *')) do
 
 					-- Is ending check and handling
 					if chapter_range.end_patterns then
-						chapter.is_end_only = false
 						for _, end_pattern in ipairs(chapter_range.end_patterns) do
 							if lowercase_title:find(end_pattern) then
 								if current_range == nil and uses_bof and not bof_used then
@@ -3193,8 +3192,8 @@ for _, definition in ipairs(split(options.chapter_ranges, ' *,+ *')) do
 								end
 								if current_range ~= nil then
 									end_range(chapter)
+									chapter.is_end_only = end_pattern ~= '.*'
 								end
-								chapter.is_end_only = end_pattern ~= '.*'
 								break
 							end
 						end
@@ -3202,7 +3201,7 @@ for _, definition in ipairs(split(options.chapter_ranges, ' *,+ *')) do
 
 					-- Is start check and handling
 					for _, start_pattern in ipairs(chapter_range.start_patterns) do
-						if lowercase_title:find(start_pattern) then
+						if lowercase_title:find(start_pattern) and current_range == nil then
 							start_range(chapter)
 							chapter.is_end_only = false
 							break
```
Which seems like an improvement in general and maybe worth upstreaming, but it still doesn't properly deal with that mess.

Maybe we should consolidate chapters that are less then  a second apart into a single one? Then every chapter would need an array of titles instead of a single one. That might improve the situation with the hovered chapters, which currently can't deal with that situation at all.

For debugging [test.zip](https://github.com/tomasklaen/uosc/files/9540598/test.zip) might be of help.

**Comment by tomasklaen** [comment]
Yeah I think we should just scrap the whole `chapter_ranges` config, and do it all internally. Looking at the chapters example you posted, it'd be trivial to make it nice internally, especially since they are nicely IDd, but almost impossible and excruciatingly awkward with `chapter_ranges` config.

Does anyone have any other use case for chapter ranges apart of:
- sponsor blocks
- anime openings & endings
- movie/show intros/credits

?

---

Btw, the menus refactoring turned into a whole element system refactoring, and I'm tweaking like 80% of the codebase, so hold off with making PRs or you might have to rewrite them. Should be done today, or monday at worst.

**Reply by OmiChrome** [reply]
anime op/ed  i don't see any way  of seeeing them working? 

idk how it works ?  :(

**Reply by christoph-heinrich** [reply]
> idk how it works ? :(

The video needs to have chapters with appropriate names.
![Screenshot_20220910_215831](https://user-images.githubusercontent.com/8932183/189499923-80c7d3ec-49c9-4a7a-83a1-0417ebdea1c6.png)
![Screenshot_20220910_220200](https://user-images.githubusercontent.com/8932183/189500015-25616543-5c89-4fe0-90a1-f986cf01af99.png)
